### PR TITLE
use tuna 🐟  to generate visual/interactive performance profiles

### PR DESCRIPTION
**Pull request by erwanp**
### Description

use [tuna](https://github.com/nschloe/tuna) to generate visual/interactive performance profiles; and the update the console profiler:

Syntax : 

```
s = calc_spectrum(...)
s.generate_perf_profile()
```

Typical output (opens a local web page) : 

image

Console profiler:

```
s = calc_spectrum(...)
s.print_perf_profile()
```

Typical output : 

```
        spectrum_calculation      0.844s ████████████████
            check_line_databank              0.001s
            check_non_eq_param               0.070s █
            fetch_energy_5                   0.027s
            calc_weight_trans                0.016s
            reinitialize                     0.003s
                copy_database                    0.000s
                memory_usage_warning             0.003s
                reset_population                 0.000s
            calc_noneq_population            0.066s █
                part_function                    0.056s █
                population                       0.010s
            scaled_non_eq_linestrength       0.004s
                map_part_func                    0.001s
                corrected_population_se          0.003s
            calc_emission_integral           0.008s
            applied_linestrength_cutoff      0.003s
            calc_lineshift                   0.001s
            calc_hwhm                        0.008s
            generate_wavenumber_arrays       0.002s
            calc_line_broadening             0.668s ████████████
                precompute_DLM_lineshapes        0.020s
                DLM_Initialized_vectors          0.000s
                DLM_closest_matching_line        0.001s
                DLM_Distribute_lines             0.002s
                DLM_convolve                     0.304s █████
            calc_other_spectral_quan         0.005s
            generate_spectrum_obj            0.000s
```

### Implementation

- [x] Proof of concept
- [x] Test
- [x] update "profiler" format consistently : @anandxkumar see below 
- [x] mention in docs/performances

@anandxkumar this expects the `conditions["profiler"]` dict to be of the format:  
```python
prof = {'scaled_eq_linestrength': 0.02393484115600586, 'calc_lineshift': 0.002992868423461914, 'calc_hwhm': 0.04986977577209473, 'precompute_DLM_lineshapes': 0.06680488586425781, 'calc_line_broadening': {'value': 1.423218011856079, 'DLM_Initialized_vectors': 0.0, 'DLM_closest_matching_line': 0.048909902572631836, 'DLM_Distribute_lines': 0.052854061126708984, 'DLM_convolve': 1.2426400184631348}, 'calc_other_spectral_quan': 0.03287649154663086, 'spectrum_calc_before_obj': 1.560823678970337, 'value': 1.560823678970337}
```
i.e. (more human-readable with `print(yaml.dump(prof, allow_unicode=True, default_flow_style=False))`
```
calc_hwhm: 0.04986977577209473
calc_line_broadening:
  DLM_Distribute_lines: 0.052854061126708984
  DLM_Initialized_vectors: 0.0
  DLM_closest_matching_line: 0.048909902572631836
  DLM_convolve: 1.2426400184631348
  value: 1.423218011856079              👈  notice the "value"
calc_lineshift: 0.002992868423461914
calc_other_spectral_quan: 0.03287649154663086
precompute_DLM_lineshapes: 0.06680488586425781
scaled_eq_linestrength: 0.02393484115600586
spectrum_calc_before_obj: 1.560823678970337
value: 1.560823678970337          👈  notice the "value"
```

This PR generates the corresponding Python Tree. 

I suggest that you use this PR to also close #332 ; 

- [x] maybe also add a Spectrum method using the `yaml.dump` function above to nicely print the profiler dictionary, adding % etc ?

### CODE REVIEW

File: radis/spectrum/spectrum.py [hunk 1]
```diff
@@ -3575,6 +3575,28 @@ def normalize(
...
+        .. note:
+            requires a `profiler` key with in Spectrum.conditions
+
+        Examples
+        --------
+        ::
+            s = calc_spectrum(...)
+            s.generate_perf_profile()
+
+        See typical output in https://github.com/radis/radis/pull/324
```
Comment (erwanp) [inline radis/spectrum/spectrum.py:3590]: should be 325, actually.

File: radis/spectrum/utils.py [hunk 2]
```diff
@@ -383,3 +385,140 @@ def split_and_plot_by_parts(w, I, *args, **kwargs):
...
+    return new_dict
+
+
+#%%
+
+
+def generate_perf_profile(s):
+    """Visual/interactive performance profile
+
+    See typical output in https://github.com/radis/radis/pull/324
```
Comment (erwanp) [inline radis/spectrum/utils.py:421]: should be 325

File: radis/test/spectrum/test_utils.py [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+def test_perf_profile(*args, **kwargs):
+    """Test visual/interactive performance profile
+
+    Examples
+    --------
+    ::
+        s = calc_spectrum(...)
+        s.generate_perf_profile()
+
+    See output in https://github.com/radis/radis/pull/324
```
Comment (erwanp) [inline radis/test/spectrum/test_utils.py:18]: should be 325.

File: radis/spectrum/spectrum.py [hunk 4]
```diff
@@ -3597,6 +3575,117 @@ def normalize(
...
+
+    def print_perf_profile(self, number_format="{:.3f}"):
+        """Prints Profiler output dictionary in a structured manner.
+
+        Example
+        -------
+        ::
+            Spectrum.print_perf_profile()
+
+            # output >>
```
Comment (erwanp) [inline radis/spectrum/spectrum.py:3588]: @anandxkumar I updated how the profiler is printed. Here for an example. It shows more explicitly what takes a lot of  time.  Can you review it also work on your side?

Comment (anandxkumar) [reply radis/spectrum/spectrum.py:3588]: I pulled the latest code of this PR, on using `Spectrum.print_perf_profile()` prints the profiler correctly, the only issue I was having on my system was the symbol seemed unidentified by the terminal or even Spyder console. I was having the following output:
```
        spectrum_calculation  0.039s â–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆâ–ˆ
        check_line_databank          0.001s 
            reinitialize                 0.003s â–ˆ
            copy_database                0.000s 
            memory_usage_warning         0.003s â–ˆ
            reset_population             0.000s 
```

Same issue on running the new code on the latest develop branch.

Comment (erwanp) [reply radis/spectrum/spectrum.py:3588]: All fixed ; have a look ! I also updated the docs extensively.

**Comment by anandxkumar** [comment]
@erwanp  Indentation Error:
```

sf = SpectrumFactory(wavenum_min=2900,
                     wavenum_max=4200,
                     molecule='OH',
                     broadening_max_width=10,  # cm-1
                     medium='vacuum',
                     verbose=3,    # more for more details
                     pressure = 10,
                     wstep = 0.01,
                     )
sf.warnings={"AccuracyError":"ignore",
             "MissingSelfBroadeningWarning":"ignore"}
sf.fetch_databank('hitemp')
s1 = sf.eq_spectrum(Tgas=300, path_length=1)
s1.print_perf_profile()
```

Output:
```
        spectrum_calculation  0.646s ████████████████
        check_line_databank         0.000s 
            reinitialize                0.008s 
            copy_database               0.001s 
            memory_usage_warning        0.001s 
            reset_population            0.000s 
        scaled_eq_linestrength      0.003s 
        calc_lineshift              0.001s 
        calc_hwhm                   0.004s 
        generate_wavenumber_arrays  0.003s 
            calc_line_broadening        0.618s ███████████████
            precompute_DLM_lineshapes   0.024s 
            DLM_Initialized_vectors     0.000s 
            DLM_closest_matching_line   0.001s 
            DLM_Distribute_lines        0.001s 
            DLM_convolve                0.591s ██████████████
        calc_other_spectral_quan    0.006s 
        generate_spectrum_obj       0.001s 
 ```
 
`reinitialize` and `calc_line_broadening` should be at the previous level indentation.

**Comment by erwanp** [comment]
Good catch. Indentation fixed: 
```
        spectrum_calculation      0.844s ████████████████
            check_line_databank              0.001s
            check_non_eq_param               0.070s █
            fetch_energy_5                   0.027s
            calc_weight_trans                0.016s
            reinitialize                     0.003s
                copy_database                    0.000s
                memory_usage_warning             0.003s
                reset_population                 0.000s
            calc_noneq_population            0.066s █
                part_function                    0.056s █
                population                       0.010s
            scaled_non_eq_linestrength       0.004s
                map_part_func                    0.001s
                corrected_population_se          0.003s
            calc_emission_integral           0.008s
            applied_linestrength_cutoff      0.003s
            calc_lineshift                   0.001s
            calc_hwhm                        0.008s
            generate_wavenumber_arrays       0.002s
            calc_line_broadening             0.668s ████████████
                precompute_DLM_lineshapes        0.020s
                DLM_Initialized_vectors          0.000s
                DLM_closest_matching_line        0.001s
                DLM_Distribute_lines             0.002s
                DLM_convolve                     0.304s █████
            calc_other_spectral_quan         0.005s
            generate_spectrum_obj            0.000s
```

**Review by anandxkumar — APPROVED** [review]

**Pull request merged**